src/gadgets/CensorshipResistanceGadget.sol

```diff
@@ -0,0 +1,84 @@
+pragma solidity ^0.8.19;
+
+import {AttestationGadget} from "src/gadgets/AttestationGadget.sol";
+
+import {IAndromeda} from "src/IAndromeda.sol";
+
+abstract contract CensorshipResistanceGadget is AttestationGadget {
+    function Suave() internal virtual returns (IAndromeda);
+
+    // require explicit onchain challenge on kettle restart
+    mapping(bytes32 => bool) private cr_challenges;
+
+    function offchain_restart_challenge() public returns (bytes32 challenge, bytes memory attestation) {
+        require(!cr_challenges[challenge]); /* already challenged */
+        challenge = _get_local_challenge(); /* do not override previous local challenge to prevent dos */
+        if (challenge == bytes32(0)) {
+            challenge = Suave().localRandom();
+            _set_local_challenge(challenge);
+        }
+        attestation = offchain_attest(this.onchain_restart_challenge.selector, abi.encodePacked(challenge));
+    }
+
+    function onchain_restart_challenge(bytes32 challenge, bytes memory attestation)
+        public
+        onchain_verify(abi.encodePacked(challenge), attestation)
+    {
+        cr_challenges[challenge] = true;
+    }
+    // epoch as seen in volatile memory of the kettle (can be delayed if kettle operator censors calls)
+
+    function _get_local_challenge() private returns (bytes32 challenge) {
+        challenge = Suave().volatileGet("local_challenge");
+    }
+
+    function _set_local_challenge(bytes32 challenge) private {
+        Suave().volatileSet("local_challenge", challenge);
+    }
+
+    // epoch as seen onchain (can be delayed if chain is censored)
+    uint256 public cr_epoch;
+
+    // epoch as seen in volatile memory of the kettle (can be delayed if kettle operator censors calls)
+    function _get_local_epoch() private returns (uint256 epoch) {
+        epoch = uint256(Suave().volatileGet("local_epoch"));
+    }
+
+    function _set_local_epoch(uint256 epoch) private {
+        Suave().volatileSet(("local_epoch"), bytes32(epoch));
+    }
+
+    function bump_cr_epoch(uint256 epoch) internal {
+        require(cr_epoch == epoch);
+        cr_epoch = epoch + 1;
+    }
+
+    // add the following modifiers to your calls to make them censorship resistant
```

Comment:
Cool, delighted to see a censorship resistance queue, like Ethereum Alarm Clock but for kettle apps. This looks like a nice way to do it with a modifier. I didn't follow how it works yet though, I was expecting a queue I guess